include/libpmemobj++/array.hpp

```diff
@@ -0,0 +1,724 @@
+/*
+ * Copyright 2018, Intel Corporation
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * Array container based on std::array
+ */
+
+#ifndef PMEMOBJ_ARRAY_HPP
+#define PMEMOBJ_ARRAY_HPP
+
+#include <algorithm>
+#include <functional>
+
+#include "libpmemobj++/array_iterator.hpp"
+#include "libpmemobj++/detail/common.hpp"
+#include "libpmemobj++/persistent_ptr.hpp"
+#include "libpmemobj++/pext.hpp"
+#include "libpmemobj++/slice.hpp"
+#include "libpmemobj.h"
+
+namespace pmem
+{
+
+namespace obj
+{
+
+/**
+ * pmem::obj::array - persistent container based on std::array
+ *
+ * All methods which allow write access to specific element add this element
+ * to a transaction.
+ *
+ * All methods which return non-const pointer to raw data add entire array
+ * to a transaction.
+ *
+ * When a non-const iterator is returned it adds part of the array
+ * to a transaction while traversing.
+ */
+template <typename T, std::size_t N>
+struct array {
+	template <typename Y, std::size_t M>
+	struct standard_array_traits {
+		using type = Y[N];
+	};
+
+	/* zero-sized array support */
+	template <typename Y>
+	struct standard_array_traits<Y, 0> {
+		struct type {
+			/* clang-format off */
+			constexpr operator Y*() const
+			{
+				return nullptr;
+			}
+			/* clang-format on */
+
+			constexpr Y &operator[](std::size_t) const
+			{
+				return *static_cast<Y *>(nullptr);
```

Comment:
this is UB AFAIK - references can't be NULL. Shouldn't any access on a zero-sized array throw out of range an exception?